monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/CombineLatestListObservable.scala

```diff
@@ -0,0 +1,163 @@
+  extends Observable[R] {
+
+  def unsafeSubscribeFn(out: Subscriber[R]): Cancelable = {
+    import out.scheduler
+
+    val numberOfObservables = obss.size
+
+    val lock = new AnyRef
+    var isDone = false
+
+    // NOTE: We use arrays and other mutable structures here to be as performant as possible.
+
+    // MUST BE synchronized by `lock`
+    val observables: Array[(Observable[A], Int)] = new Array(numberOfObservables)
+    obss.zipWithIndex.foreach { obsAndIndex =>
+      observables(obsAndIndex._2) = obsAndIndex
+    }
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+    // MUST BE synchronized by `lock`
+    val elems: mutable.IndexedSeq[A] = mutable.IndexedSeq.fill(numberOfObservables)(null.asInstanceOf[A])
+    // MUST BE synchronized by `lock`
+    val hasElems: Array[Boolean] = Array.fill(numberOfObservables)(false)
+
+    var hasElemsCount: Int = 0
+
+    // MUST BE synchronized by `lock`
+    var completedCount = 0
+
+    // MUST BE synchronized by `lock`
+    def rawOnNext(as: Seq[A]): Future[Ack] =
+      if (isDone) Stop
+      else {
+        var streamError = true
+        try {
+          val c = f(as)
+          streamError = false
+          out.onNext(c)
+        } catch {
+          case NonFatal(ex) if streamError =>
+            isDone = true
+            out.onError(ex)
+            Stop
+        }
+      }
+
+    // MUST BE synchronized by `lock`
+    def signalOnNext(as: Seq[A]): Future[Ack] = {
+      lastAck = lastAck match {
+        case Continue => rawOnNext(as)
+        case Stop => Stop
+        case async =>
+          async.flatMap {
+            // async execution, we have to re-sync
+            case Continue => lock.synchronized(rawOnNext(as))
+            case Stop => Stop
+          }
+      }
+
+      lastAck
+    }
+
+    def signalOnError(ex: Throwable): Unit = lock.synchronized {
+      if (!isDone) {
+        isDone = true
+        out.onError(ex)
+        lastAck = Stop
+      }
+    }
+
+    def signalOnComplete(): Unit = lock.synchronized {
+      completedCount += 1
+
+      if (completedCount == numberOfObservables && !isDone) {
+        lastAck match {
+          case Continue =>
+            isDone = true
+            out.onComplete()
+          case Stop =>
+            () // do nothing
+          case async =>
+            async.onComplete {
+              case Success(Continue) =>
+                lock.synchronized {
+                  if (!isDone) {
+                    isDone = true
+                    out.onComplete()
+                  }
+                }
+              case _ =>
+                () // do nothing
+            }
+        }
+
+        lastAck = Stop
+      }
+    }
+
+    val composite = CompositeCancelable()
+
+    observables.foreach { case (obs, index) =>
+      composite += obs.unsafeSubscribeFn(new Subscriber[A] {
+        implicit val scheduler: Scheduler = out.scheduler
+
```

Comment:
Correct me if I wrong, but why do we need these tuples in `observables`? It seems that `index` is only used here, so one should be able to drop these allocations just by doing:
```var i = 0
observables.foreach { obs =>
  val index = i // close over proper index
  i += 1
```
and thus you don't need to reallocate collection at all (can just use original `obss` coming to you)

Reply:
Or
```
observables.iterator.zipWithIndex.map { case (obs, index) =>
    ???
}}}
```

Reply:
@jakoschiko that is indeed an option, but it's only halfway - you skip the allocation of an extra collection, but not the tuples for each instance.